idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddExclExclCallFixFactories.kt

```diff
@@ -0,0 +1,121 @@
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtFunctionSymbol
+import org.jetbrains.kotlin.idea.frontend.api.types.KtType
+import org.jetbrains.kotlin.idea.quickfix.AddExclExclCallFix
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+object AddExclExclCallFixFactories {
+    private fun PsiElement?.asFix(hasImplicitReceiver: Boolean = false) =
+        this?.let { listOf(AddExclExclCallFix(it, hasImplicitReceiver)) } ?: emptyList()
+
+    val unsafeCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    val unsafeInfixCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeInfixCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    val unsafeOperatorCallFactory = diagnosticFixFactory<KtFirDiagnostic.UnsafeOperatorCall> { diagnostic ->
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    private fun KtAnalysisSession.getFixForUnsafeCall(psi: PsiElement): List<IntentionAction> {
+        val (target, hasImplicitReceiver) = when (psi) {
+            // `foo.bar` -> `foo!!.bar`
+            is KtDotQualifiedExpression -> psi.receiverExpression to false
+
+            // `foo[bar]` -> `foo!![bar]`
+            is KtArrayAccessExpression -> psi.arrayExpression to false
+
+            is KtCallableReferenceExpression -> psi.lhs.let { lhs ->
+                if (lhs != null) {
+                    // `foo::bar` -> `foo!!::bar`
+                    lhs to false
+                } else {
+                    // `::bar -> this!!::bar`
+                    psi to true
+                }
+            }
+
+            // `bar` -> `this!!.bar`
+            is KtNameReferenceExpression -> psi to true
+
+            // `bar()` -> `this!!.bar()`
+            is KtCallExpression -> psi to true
+
+            // `-foo` -> `-foo!!`
+            // NOTE: Unsafe unary operator call is reported as UNSAFE_CALL, _not_ UNSAFE_OPERATOR_CALL
+            is KtUnaryExpression -> psi.baseExpression to false
+
+            is KtBinaryExpression -> {
+                val receiver = if (KtPsiUtil.isInOrNotInOperation(psi)) {
+                    // `bar in foo` -> `bar in foo!!`
+                    psi.right
+                } else {
+                    // `foo + bar` -> `foo!! + bar` OR `foo infixFun bar` -> `foo!! infixFun bar`
+                    psi.left
+                }
+                receiver to false
+            }
+
+            // UNSAFE_INFIX_CALL/UNSAFE_OPERATOR_CALL on KtBinaryExpression is reported on the child KtOperationReferenceExpression
+            is KtOperationReferenceExpression -> return getFixForUnsafeCall(psi.parent)
+
+            else -> return emptyList()
+        }
+
+        return target.asFix(hasImplicitReceiver = hasImplicitReceiver)
+    }
+
+    // TODO: Move these factories to separate file when there are more *TYPE_MISMATCH fixes available.
```

Comment:
Can you move it to a separate file in this PR? I will be adding wrap with safe let call quickfix to type mismatch soon and it would be easier to collabrate in case you will add more type mismatch quickfixes in the near future.